lib/Dialect/SV/Transforms/RTLMemSimImpl.cpp

```diff
@@ -0,0 +1,203 @@
+//===- RTLMemSimImpl.cpp - RTL Memory Implementation Pass -----------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This transformation pass converts gemerated FIRRTL memory modules to
+// simulation models.
+//
+//===----------------------------------------------------------------------===//
+
+#include "SVPassDetail.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "circt/Support/ImplicitLocOpBuilder.h"
+
+using namespace circt;
+
+//===----------------------------------------------------------------------===//
+// StubExternalModules Pass
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct RTLMemSimImplPass : public sv::RTLMemSimImplBase<RTLMemSimImplPass> {
+  void runOnOperation() override;
+
+private:
+  void generateMemory(rtl::RTLModuleOp op, rtl::RTLModuleGeneratedOp refOp);
+};
+
+struct FirMemory {
+  size_t numReadPorts;
+  size_t numWritePorts;
+  size_t numReadWritePorts;
+  size_t dataWidth;
+  size_t depth;
+  size_t readLatency;
+  size_t writeLatency;
+  size_t readUnderWrite;
+};
+} // end anonymous namespace
+
+static FirMemory analyzeMemOp(rtl::RTLModuleGeneratedOp op) {
+  FirMemory mem;
+  mem.depth = op->getAttrOfType<IntegerAttr>("depth").getInt();
+  mem.numReadPorts = op->getAttrOfType<IntegerAttr>("numReadPorts").getUInt();
+  mem.numWritePorts = op->getAttrOfType<IntegerAttr>("numWritePorts").getUInt();
+  mem.numReadWritePorts =
+      op->getAttrOfType<IntegerAttr>("numReadWritePorts").getUInt();
+  mem.readLatency = op->getAttrOfType<IntegerAttr>("readLatency").getUInt();
+  mem.writeLatency = op->getAttrOfType<IntegerAttr>("writeLatency").getUInt();
+  mem.dataWidth = op->getAttrOfType<IntegerAttr>("width").getUInt();
+  mem.readUnderWrite =
+      op->getAttrOfType<IntegerAttr>("readUnderWrite").getUInt();
+  return mem;
+};
+
+static Value addPipelineStages(ImplicitLocOpBuilder &b, size_t stages,
```

Comment:
Nothing to do about this, but for the record, this seems like another place we could use #587.

Reply:
Yup, though I was please with how compact it wound up being.

Reply:
It's doing what scala firrtl is doing and generating a register per value per delay cycle, which is ugly, but can be cleaned up later.

Reply:
Yep totally agree.